File: agentic-loops/react_loop.go
// Production ReAct Loop — Go
// Reason → Act → Observe, repeat until finish() called.
//
// go get github.com/anthropics/anthropic-sdk-go
// export ANTHROPIC_API_KEY=sk-ant-...
// go run react_loop.go "your task here"

package main

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"os/exec"
	"regexp"
	"strings"
	"time"

	anthropic "github.com/anthropics/anthropic-sdk-go"
	"github.com/anthropics/anthropic-sdk-go/option"
)

// ── Trace types ───────────────────────────────────────────────────────────────

type StepType string

const (
	StepThought     StepType = "thought"
	StepAction      StepType = "action"
	StepObservation StepType = "observation"
	StepAnswer      StepType = "answer"
)

type Step struct {
	Type    StepType       `json:"type"`
	Content string         `json:"content,omitempty"`
	Tool    string         `json:"tool,omitempty"`
	Input   map[string]any `json:"input,omitempty"`
	Error   bool           `json:"error,omitempty"`
}

type Trace struct {
	Steps       []Step        `json:"steps"`
	TotalTokens int64         `json:"total_tokens"`
	Iterations  int           `json:"iterations"`
	Elapsed     time.Duration `json:"elapsed_ms"`
}

// ── Tool registry ─────────────────────────────────────────────────────────────

type ToolFunc func(input map[string]any) (string, error)

var (
	toolRegistry = map[string]ToolFunc{}
	toolSchemas  []anthropic.ToolParam
)

func registerTool(name, description string, schema map[string]any, fn ToolFunc) {
	toolRegistry[name] = fn
	toolSchemas = append(toolSchemas, anthropic.ToolParam{
		Name:        name,
		Description: anthropic.String(description),
		InputSchema: anthropic.ToolInputSchemaParam{
			Properties: anthropic.F[any](schema["properties"]),
		},
	})
}

// ── Built-in tools ────────────────────────────────────────────────────────────

func init() {
	registerTool("think", "Reason step by step before acting.",
		map[string]any{
			"type": "object",
			"properties": map[string]any{
				"reasoning": map[string]any{"type": "string"},
			},
			"required": []string{"reasoning"},
		},
		func(input map[string]any) (string, error) {
			return "OK", nil
		},
	)

	registerTool("search", "Search the web for current information.",
		map[string]any{
			"type": "object",
			"properties": map[string]any{
				"query": map[string]any{"type": "string"},
			},
			"required": []string{"query"},
		},
		func(input map[string]any) (string, error) {
			// Replace with Brave, Tavily, SerpAPI, etc.
			return "", fmt.Errorf("search not wired up. query: %v", input["query"])
		},
	)

	registerTool("fetch_url", "Fetch the text content of a URL.",
		map[string]any{
			"type": "object",
			"properties": map[string]any{
				"url": map[string]any{"type": "string"},
			},
			"required": []string{"url"},
		},
		func(input map[string]any) (string, error) {
			url, _ := input["url"].(string)
			resp, err := http.Get(url)
			if err != nil {
				return "", err
			}
			defer resp.Body.Close()
			raw, err := io.ReadAll(io.LimitReader(resp.Body, 1<<20))
			if err != nil {
				return "", err
			}
			// Strip HTML
			re := regexp.MustCompile(`<[^>]+>`)
			clean := re.ReplaceAllString(string(raw), " ")
			clean = strings.Join(strings.Fields(clean), " ")
			if len(clean) > 8000 {
				clean = clean[:8000]
			}
			return clean, nil
		},
	)

	registerTool("run_bash", "Execute a bash command. Returns stdout/stderr. Timeout 15s.",
		map[string]any{
			"type": "object",
			"properties": map[string]any{
				"command": map[string]any{"type": "string"},
			},
			"required": []string{"command"},
		},
		func(input map[string]any) (string, error) {
			command, _ := input["command"].(string)
			ctx, cancel := context.WithTimeout(context.Background(), 15*time.Second)
			defer cancel()
			cmd := exec.CommandContext(ctx, "bash", "-c", command)
			var out bytes.Buffer
			cmd.Stdout = &out
			cmd.Stderr = &out
			_ = cmd.Run()
			result := strings.TrimSpace(out.String())
			if result == "" {
				return "(no output)", nil
			}
			if len(result) > 8000 {
				result = result[:8000]
			}
			return result, nil
		},
	)

	registerTool("read_file", "Read a file from disk.",
		map[string]any{
			"type": "object",
			"properties": map[string]any{
				"path": map[string]any{"type": "string"},
			},
			"required": []string{"path"},
		},
		func(input map[string]any) (string, error) {
			path, _ := input["path"].(string)
			data, err := os.ReadFile(path)
			return string(data), err
		},
	)

	registerTool("write_file", "Write content to a file on disk.",
		map[string]any{
			"type": "object",
			"properties": map[string]any{
				"path":    map[string]any{"type": "string"},
				"content": map[string]any{"type": "string"},
			},
			"required": []string{"path", "content"},
		},
		func(input map[string]any) (string, error) {
			path, _ := input["path"].(string)
			content, _ := input["content"].(string)
			err := os.WriteFile(path, []byte(content), 0644)
			if err != nil {
				return "", err
			}
			return fmt.Sprintf("Written %d bytes to %s", len(content), path), nil
		},
	)

	registerTool("finish", "Call when you have the final answer. Ends the loop.",
		map[string]any{
			"type": "object",
			"properties": map[string]any{
				"answer": map[string]any{"type": "string"},
			},
			"required": []string{"answer"},
		},
		func(input map[string]any) (string, error) {
			answer, _ := input["answer"].(string)
			return answer, nil
		},
	)
}

// ── ReAct system prompt ───────────────────────────────────────────────────────

const reactSystem = `You are an autonomous agent operating in a ReAct loop (Reason + Act + Observe).

For every task:
1. Use the think tool to reason about what to do next before acting.
2. Call the appropriate tool to act.
3. Observe the result and reason again.
4. Repeat until you have a complete, verified answer.
5. Call finish with your final answer when done.

Rules:
- Always think before acting. Never skip the think step.
- If a tool errors, reason about why and try a different approach.
- Do not guess. If unsure, search or fetch.
- Be thorough. Do not call finish until the task is fully complete.`

// ── Core ReAct engine ─────────────────────────────────────────────────────────

type ReactOptions struct {
	SystemExtra   string
	MaxIterations int
	OnStep        func(Step)
}

func RunReact(ctx context.Context, task string, opts ReactOptions) (string, *Trace, error) {
	if opts.MaxIterations == 0 {
		opts.MaxIterations = 30
	}

	client := anthropic.NewClient(option.WithAPIKey(os.Getenv("ANTHROPIC_API_KEY")))
	system := reactSystem
	if opts.SystemExtra != "" {
		system += "\n\n" + opts.SystemExtra
	}

	messages := []anthropic.MessageParam{
		anthropic.NewUserMessage(anthropic.NewTextBlock(task)),
	}

	trace := &Trace{}
	start := time.Now()
	fmt.Fprintf(os.Stderr, "[react] START: %s\n", truncate(task, 120))

	for i := 0; i < opts.MaxIterations; i++ {
		trace.Iterations = i + 1
		fmt.Fprintf(os.Stderr, "[react] iteration %d\n", i+1)

		response, err := client.Messages.New(ctx, anthropic.MessageNewParams{
			Model:     anthropic.F(anthropic.ModelClaudeOpus46),
			MaxTokens: anthropic.F(int64(4096)),
			System:    anthropic.F([]anthropic.TextBlockParam{{Type: "text", Text: system}}),
			Tools:     anthropic.F(toolSchemas),
			Messages:  anthropic.F(messages),
		})
		if err != nil {
			return "", trace, fmt.Errorf("API error: %w", err)
		}

		trace.TotalTokens += response.Usage.InputTokens + response.Usage.OutputTokens
		messages = append(messages, response.ToParam())

		if response.StopReason == anthropic.StopReasonEndTurn {
			for _, block := range response.Content {
				if block.Type == anthropic.ContentBlockTypeText && strings.TrimSpace(block.Text) != "" {
					answer := strings.TrimSpace(block.Text)
					trace.Steps = append(trace.Steps, Step{Type: StepAnswer, Content: answer})
					trace.Elapsed = time.Since(start)
					return answer, trace, nil
				}
			}
		}

		if response.StopReason == anthropic.StopReasonToolUse {
			var toolResults []anthropic.ContentBlockParamUnion

			for _, block := range response.Content {
				if block.Type != anthropic.ContentBlockTypeToolUse {
					continue
				}

				name := block.Name
				var input map[string]any
				_ = json.Unmarshal([]byte(block.Input.(string)), &input)
				if input == nil {
					if raw, ok := block.Input.(map[string]any); ok {
						input = raw
					}
				}

				// Think
				if name == "think" {
					reasoning, _ := input["reasoning"].(string)
					step := Step{Type: StepThought, Content: reasoning}
					trace.Steps = append(trace.Steps, step)
					if opts.OnStep != nil {
						opts.OnStep(step)
					}
					fmt.Fprintf(os.Stderr, "[react] 💭 %s\n", truncate(reasoning, 200))
					toolResults = append(toolResults, anthropic.NewToolResultBlock(block.ID, "OK", false))
					continue
				}

				// Finish
				if name == "finish" {
					answer, _ := input["answer"].(string)
					trace.Steps = append(trace.Steps, Step{Type: StepAnswer, Content: answer})
					trace.Elapsed = time.Since(start)
					fmt.Fprintf(os.Stderr, "[react] DONE — %d iters, %d tokens, %v\n",
						trace.Iterations, trace.TotalTokens, trace.Elapsed.Round(time.Millisecond))
					return answer, trace, nil
				}

				// Regular action
				step := Step{Type: StepAction, Tool: name, Input: input}
				trace.Steps = append(trace.Steps, step)
				if opts.OnStep != nil {
					opts.OnStep(step)
				}
				inJSON, _ := json.Marshal(input)
				fmt.Fprintf(os.Stderr, "[react] ⚡ %s(%s)\n", name, truncate(string(inJSON), 200))

				fn, ok := toolRegistry[name]
				var result string
				isError := false
				if !ok {
					result = fmt.Sprintf("Tool '%s' not registered", name)
					isError = true
				} else {
					result, err = fn(input)
					if err != nil {
						result = fmt.Sprintf("Tool error: %v", err)
						isError = true
						fmt.Fprintf(os.Stderr, "[react] ❌ %s: %s\n", name, result)
					}
				}

				obs := Step{Type: StepObservation, Content: result, Error: isError}
				trace.Steps = append(trace.Steps, obs)
				if opts.OnStep != nil {
					opts.OnStep(obs)
				}
				fmt.Fprintf(os.Stderr, "[react] 👁 %s\n", truncate(result, 200))
				toolResults = append(toolResults, anthropic.NewToolResultBlock(block.ID, result, isError))
			}

			messages = append(messages, anthropic.NewUserMessage(toolResults...))
			continue
		}

		fmt.Fprintf(os.Stderr, "[react] unexpected stop_reason: %s\n", response.StopReason)
		break
	}

	trace.Elapsed = time.Since(start)
	return "", trace, fmt.Errorf("ReAct loop did not finish within %d iterations (tokens: %d)",
		opts.MaxIterations, trace.TotalTokens)
}

func truncate(s string, n int) string {
	if len(s) <= n {
		return s
	}
	return s[:n] + "..."
}

// ── CLI entry point ───────────────────────────────────────────────────────────

func main() {
	task := strings.Join(os.Args[1:], " ")
	if task == "" {
		task = "Calculate the 47th Fibonacci number using bash, then explain the result."
	}

	answer, trace, err := RunReact(context.Background(), task, ReactOptions{
		OnStep: func(step Step) {
			switch step.Type {
			case StepThought:
				fmt.Printf("\n💭 THOUGHT: %s\n", truncate(step.Content, 300))
			case StepAction:
				inp, _ := json.Marshal(step.Input)
				fmt.Printf("\n⚡ ACTION: %s(%s)\n", step.Tool, truncate(string(inp), 200))
			case StepObservation:
				marker := "👁"
				if step.Error {
					marker = "❌"
				}
				fmt.Printf("\n%s OBSERVE: %s\n", marker, truncate(step.Content, 300))
			}
		},
	})

	if err != nil {
		fmt.Fprintf(os.Stderr, "FATAL: %v\n", err)
		os.Exit(1)
	}

	fmt.Println("\n" + strings.Repeat("=", 60))
	fmt.Println("FINAL ANSWER:")
	fmt.Println(answer)
	fmt.Println(strings.Repeat("=", 60))
	fmt.Printf("Iterations: %d | Tokens: %d | Time: %v\n",
		trace.Iterations, trace.TotalTokens, trace.Elapsed.Round(time.Millisecond))

	out, _ := json.MarshalIndent(trace.Steps, "", "  ")
	fmt.Println("\nFull trace:")
	fmt.Println(string(out))
}

File: agentic-loops/react_loop.rs
//! Production ReAct Loop — Rust
//! Reason → Act → Observe, repeat until finish() called.
//!
//! Cargo.toml:
//! [dependencies]
//! reqwest = { version = "0.12", features = ["json", "blocking"] }
//! tokio = { version = "1", features = ["full"] }
//! serde = { version = "1", features = ["derive"] }
//! serde_json = "1"
//! anyhow = "1"
//!
//! export ANTHROPIC_API_KEY=sk-ant-...
//! cargo run -- "your task here"

use std::collections::HashMap;
use std::env;
use std::fs;
use std::process::Command;
use std::sync::Arc;
use std::time::Instant;

use anyhow::{anyhow, Context, Result};
use reqwest::Client;
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};

// ── Trace types ───────────────────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum Step {
    Thought { content: String },
    Action { tool: String, input: Value },
    Observation { content: String, error: bool },
    Answer { content: String },
}

#[derive(Debug, Default, Serialize)]
pub struct Trace {
    pub steps:        Vec<Step>,
    pub total_tokens: u64,
    pub iterations:   usize,
    pub elapsed_ms:   u128,
}

// ── Tool registry ─────────────────────────────────────────────────────────────

type ToolFn = Arc<dyn Fn(Value) -> Result<String> + Send + Sync>;

#[derive(Clone)]
pub struct ToolDef {
    pub name:        String,
    pub description: String,
    pub schema:      Value,
    pub func:        ToolFn,
}

pub struct ToolRegistry {
    tools: HashMap<String, ToolDef>,
}

impl ToolRegistry {
    pub fn new() -> Self {
        let mut r = Self { tools: HashMap::new() };
        r.register_defaults();
        r
    }

    pub fn register(&mut self, name: &str, description: &str, schema: Value, func: ToolFn) {
        self.tools.insert(name.to_string(), ToolDef {
            name:        name.to_string(),
            description: description.to_string(),
            schema,
            func,
        });
    }

    pub fn schemas(&self) -> Vec<Value> {
        self.tools.values().map(|t| json!({
            "name":         t.name,
            "description":  t.description,
            "input_schema": t.schema,
        })).collect()
    }

    pub fn call(&self, name: &str, input: Value) -> Result<String> {
        let def = self.tools.get(name)
            .ok_or_else(|| anyhow!("Tool '{}' not registered", name))?;
        (def.func)(input)
    }

    fn register_defaults(&mut self) {
        self.register("think",
            "Reason step by step before acting.",
            json!({
                "type": "object",
                "properties": { "reasoning": { "type": "string" } },
                "required": ["reasoning"]
            }),
            Arc::new(|_| Ok("OK".to_string())),
        );

        self.register("search",
            "Search the web for current information.",
            json!({
                "type": "object",
                "properties": { "query": { "type": "string" } },
                "required": ["query"]
            }),
            Arc::new(|input| {
                // Replace with Brave, Tavily, SerpAPI, etc.
                Err(anyhow!("Search not wired up. query: {}", input["query"]))
            }),
        );

        self.register("fetch_url",
            "Fetch the text content of a URL.",
            json!({
                "type": "object",
                "properties": { "url": { "type": "string" } },
                "required": ["url"]
            }),
            Arc::new(|input| {
                let url = input["url"].as_str().unwrap_or("").to_string();
                let body = reqwest::blocking::get(&url)
                    .context("HTTP GET failed")?
                    .text()
                    .context("Read body failed")?;
                // Strip HTML tags (simple regex-free approach)
                let mut result = String::new();
                let mut in_tag = false;
                for c in body.chars() {
                    match c {
                        '<' => in_tag = true,
                        '>' => in_tag = false,
                        _ if !in_tag => result.push(c),
                        _ => {}
                    }
                }
                let clean: String = result.split_whitespace().collect::<Vec<_>>().join(" ");
                Ok(clean.chars().take(8000).collect())
            }),
        );

        self.register("run_bash",
            "Execute a bash command. Returns stdout/stderr. Timeout 15s.",
            json!({
                "type": "object",
                "properties": { "command": { "type": "string" } },
                "required": ["command"]
            }),
            Arc::new(|input| {
                let command = input["command"].as_str().unwrap_or("").to_string();
                let output = Command::new("bash")
                    .args(["-c", &command])
                    .output()
                    .context("Failed to run bash")?;
                let stdout = String::from_utf8_lossy(&output.stdout).to_string();
                let stderr = String::from_utf8_lossy(&output.stderr).to_string();
                let combined = format!("{}{}", stdout, stderr).trim().to_string();
                if combined.is_empty() {
                    return Ok("(no output)".to_string());
                }
                Ok(combined.chars().take(8000).collect())
            }),
        );

        self.register("read_file",
            "Read a file from disk.",
            json!({
                "type": "object",
                "properties": { "path": { "type": "string" } },
                "required": ["path"]
            }),
            Arc::new(|input| {
                let path = input["path"].as_str().unwrap_or("").to_string();
                fs::read_to_string(&path).context("Read file failed")
            }),
        );

        self.register("write_file",
            "Write content to a file on disk.",
            json!({
                "type": "object",
                "properties": {
                    "path":    { "type": "string" },
                    "content": { "type": "string" }
                },
                "required": ["path", "content"]
            }),
            Arc::new(|input| {
                let path    = input["path"].as_str().unwrap_or("").to_string();
                let content = input["content"].as_str().unwrap_or("").to_string();
                let len = content.len();
                fs::write(&path, &content).context("Write file failed")?;
                Ok(format!("Written {} bytes to {}", len, path))
            }),
        );

        self.register("finish",
            "Call when you have the final answer. Ends the loop.",
            json!({
                "type": "object",
                "properties": { "answer": { "type": "string" } },
                "required": ["answer"]
            }),
            Arc::new(|input| Ok(input["answer"].as_str().unwrap_or("").to_string())),
        );
    }
}

// ── API types ─────────────────────────────────────────────────────────────────

#[derive(Deserialize, Debug)]
struct ApiResponse {
    content:     Vec<ContentBlock>,
    stop_reason: String,
    usage:       Usage,
}

#[derive(Deserialize, Debug)]
struct ContentBlock {
    #[serde(rename = "type")]
    kind:  String,
    text:  Option<String>,
    id:    Option<String>,
    name:  Option<String>,
    input: Option<Value>,
}

#[derive(Deserialize, Debug)]
struct Usage {
    input_tokens:  u64,
    output_tokens: u64,
}

// ── ReAct system prompt ───────────────────────────────────────────────────────

const REACT_SYSTEM: &str = "You are an autonomous agent operating in a ReAct loop (Reason + Act + Observe).\n\n\
For every task:\n\
1. Use the think tool to reason about what to do next before acting.\n\
2. Call the appropriate tool to act.\n\
3. Observe the result and reason again.\n\
4. Repeat until you have a complete, verified answer.\n\
5. Call finish with your final answer when done.\n\n\
Rules:\n\
- Always think before acting. Never skip the think step.\n\
- If a tool errors, reason about why and try a different approach.\n\
- Do not guess. If unsure, search or fetch.\n\
- Be thorough. Do not call finish until the task is fully complete.";

// ── Core ReAct engine ─────────────────────────────────────────────────────────

pub struct ReactOptions {
    pub system_extra:   String,
    pub max_iterations: usize,
    pub on_step:        Option<Box<dyn Fn(&Step) + Send>>,
}

impl Default for ReactOptions {
    fn default() -> Self {
        Self {
            system_extra:   String::new(),
            max_iterations: 30,
            on_step:        None,
        }
    }
}

pub async fn run_react(
    task: &str,
    registry: &ToolRegistry,
    opts: ReactOptions,
) -> Result<(String, Trace)> {
    let api_key = env::var("ANTHROPIC_API_KEY").context("ANTHROPIC_API_KEY not set")?;
    let http    = Client::builder().timeout(std::time::Duration::from_secs(60)).build()?;

    let system = if opts.system_extra.is_empty() {
        REACT_SYSTEM.to_string()
    } else {
        format!("{}\n\n{}", REACT_SYSTEM, opts.system_extra)
    };

    let mut messages: Vec<Value> = vec![json!({ "role": "user", "content": task })];
    let mut trace = Trace::default();
    let start = Instant::now();

    eprintln!("[react] START: {}", &task[..task.len().min(120)]);

    for i in 0..opts.max_iterations {
        trace.iterations = i + 1;
        eprintln!("[react] iteration {}", i + 1);

        let body = json!({
            "model":      "claude-opus-4-6",
            "max_tokens": 4096,
            "system":     system,
            "tools":      registry.schemas(),
            "messages":   messages,
        });

        let resp = http.post("https://api.anthropic.com/v1/messages")
            .header("x-api-key", &api_key)
            .header("anthropic-version", "2023-06-01")
            .header("content-type", "application/json")
            .json(&body)
            .send()
            .await
            .context("HTTP send failed")?;

        let status = resp.status();
        let text = resp.text().await?;
        if !status.is_success() {
            return Err(anyhow!("API error {}: {}", status, text));
        }

        let api: ApiResponse = serde_json::from_str(&text)
            .context("Failed to parse API response")?;

        trace.total_tokens += api.usage.input_tokens + api.usage.output_tokens;

        // Rebuild assistant content as Value array for message history
        let assistant_content: Vec<Value> = api.content.iter().map(|b| {
            let mut v = json!({ "type": b.kind });
            if let Some(t) = &b.text  { v["text"]  = json!(t); }
            if let Some(id) = &b.id   { v["id"]    = json!(id); }
            if let Some(n) = &b.name  { v["name"]  = json!(n); }
            if let Some(inp) = &b.input { v["input"] = inp.clone(); }
            v
        }).collect();
        messages.push(json!({ "role": "assistant", "content": assistant_content }));

        if api.stop_reason == "end_turn" {
            for block in &api.content {
                if block.kind == "text" {
                    if let Some(text) = &block.text {
                        let answer = text.trim().to_string();
                        trace.steps.push(Step::Answer { content: answer.clone() });
                        trace.elapsed_ms = start.elapsed().as_millis();
                        return Ok((answer, trace));
                    }
                }
            }
        }

        if api.stop_reason == "tool_use" {
            let mut tool_results: Vec<Value> = Vec::new();

            for block in &api.content {
                if block.kind != "tool_use" { continue; }
                let name  = block.name.as_deref().unwrap_or("");
                let id    = block.id.as_deref().unwrap_or("");
                let input = block.input.clone().unwrap_or(json!({}));

                // Think
                if name == "think" {
                    let reasoning = input["reasoning"].as_str().unwrap_or("").to_string();
                    let step = Step::Thought { content: reasoning.clone() };
                    if let Some(cb) = &opts.on_step { cb(&step); }
                    trace.steps.push(step);
                    eprintln!("[react] 💭 {}", &reasoning[..reasoning.len().min(200)]);
                    tool_results.push(json!({
                        "type": "tool_result", "tool_use_id": id, "content": "OK"
                    }));
                    continue;
                }

                // Finish
                if name == "finish" {
                    let answer = input["answer"].as_str().unwrap_or("").to_string();
                    trace.steps.push(Step::Answer { content: answer.clone() });
                    trace.elapsed_ms = start.elapsed().as_millis();
                    eprintln!("[react] DONE — {} iters, {} tokens, {}ms",
                        trace.iterations, trace.total_tokens, trace.elapsed_ms);
                    return Ok((answer, trace));
                }

                // Regular action
                let step = Step::Action { tool: name.to_string(), input: input.clone() };
                if let Some(cb) = &opts.on_step { cb(&step); }
                trace.steps.push(step);
                eprintln!("[react] ⚡ {}({})", name, input.to_string().chars().take(200).collect::<String>());

                let (result, is_error) = match registry.call(name, input) {
                    Ok(r)  => (r, false),
                    Err(e) => {
                        eprintln!("[react] ❌ {}: {}", name, e);
                        (format!("Tool error: {}", e), true)
                    }
                };

                let obs = Step::Observation { content: result.clone(), error: is_error };
                if let Some(cb) = &opts.on_step { cb(&obs); }
                trace.steps.push(obs);
                eprintln!("[react] 👁 {}", &result[..result.len().min(200)]);

                tool_results.push(json!({
                    "type":        "tool_result",
                    "tool_use_id": id,
                    "content":     result,
                    "is_error":    is_error,
                }));
            }

            messages.push(json!({ "role": "user", "content": tool_results }));
            continue;
        }

        eprintln!("[react] unexpected stop_reason: {}", api.stop_reason);
        break;
    }

    trace.elapsed_ms = start.elapsed().as_millis();
    Err(anyhow!(
        "ReAct loop did not finish within {} iterations (tokens: {})",
        opts.max_iterations, trace.total_tokens
    ))
}

// ── CLI entry point ───────────────────────────────────────────────────────────

#[tokio::main]
async fn main() -> Result<()> {
    let args: Vec<String> = env::args().skip(1).collect();
    let task = if args.is_empty() {
        "Calculate the 47th Fibonacci number using bash, then explain the result.".to_string()
    } else {
        args.join(" ")
    };

    let registry = ToolRegistry::new();

    let (answer, trace) = run_react(&task, &registry, ReactOptions {
        on_step: Some(Box::new(|step| match step {
            Step::Thought { content } =>
                println!("\n💭 THOUGHT: {}", &content[..content.len().min(300)]),
            Step::Action { tool, input } =>
                println!("\n⚡ ACTION: {}({})", tool, input.to_string().chars().take(200).collect::<String>()),
            Step::Observation { content, error } =>
                println!("\n{} OBSERVE: {}", if *error { "❌" } else { "👁" }, &content[..content.len().min(300)]),
            _ => {}
        })),
        ..Default::default()
    }).await?;

    println!("\n{}", "=".repeat(60));
    println!("FINAL ANSWER:\n{}", answer);
    println!("{}", "=".repeat(60));
    println!("Iterations: {} | Tokens: {} | Time: {}ms",
        trace.iterations, trace.total_tokens, trace.elapsed_ms);
    println!("\nFull trace:");
    println!("{}", serde_json::to_string_pretty(&trace.steps)?);

    Ok(())
}
